### PR TITLE
refactor(promise-kit): Rename PromiseRecord type to PromiseKit.

### PR DESCRIPTION
**Pull request by kriskowal**

### CODE REVIEW

File: packages/promise-kit/src/promiseKit.js [hunk 1]
```diff
@@ -7,14 +7,25 @@
...
  * @property {(value: ERef<T>) => void} resolve
  * @property {(reason: any) => void} reject
  * @property {Promise<T>} promise
  */
 
+/**
+ * PromiseRecord is deprecated in favor of PromiseKit.
+ *
+ * @template T
+ * @typedef {PromiseKit<T>} PromsieRecord
```
Comment (Chris-Hibbert) [inline packages/promise-kit/src/promiseKit.js]:
```suggestion
 * @typedef {PromiseKit<T>} PromiseRecord
```

And doesn't typescript allow explicit deprecation?  That's just _dicta_

Comment (kriskowal) [reply packages/promise-kit/src/promiseKit.js]: It does. I’ll check how well it interacts with tooling.

I’ve found in the past that tooling that sees _and enforces_ deprecation somewhat runs afoul of the intent of providing a path for gradual migration. I’ll check.

**Review by mhofman — COMMENTED** [review]
My main concern is when `promise-kit` is imported before the module that performs `lockdown()`, but `makePromiseKit()` is used after, the return value will just be frozen instead of hardened as might be expected (aka the promise, resolve and reject will not be frozen/hardened).

**Comment by erights** [comment]
I find the replacement of `PromiseRecord` by `PromiseKit` compelling.

On the other change, we've paid a lot of costs in the past trying to write code that could work in or out of SES. These seemed simple enough at the time, but the extra complexity caused us to back out of most of these. The only one I can recall where the costs were worth it is eventual-send/HandledPromise. Because we're proposing HandledPromise to tc39 independently of SES, a non-SES-dependent shim is worth it.

For promiseKit, what concretely motivates enabling it to be used outside SES? Did we run into a case where we want to do this ourselves? Why?

**Review by erights — CHANGES_REQUESTED** [review]
See comment at https://github.com/Agoric/agoric-sdk/pull/4201#issuecomment-999231194

**Comment by kriskowal** [comment]
Starting with xsnap, I started using async iterators as a stream abstraction. I’ve since factored out an async iterator `@endo/netstring` library that _embeds_ a copy of a `stream.js` (`queue`, `pipe` and types for `Stream`, `Reader`, and `Writer` that are async iterator/generator compatible) which in turn _embeds_ a minimal `promise-kit.js` and doesn’t depend on a hardened environment. With the work I’m doing on Endo Browser, I’ve needed to go farther, with stream libraries for native messaging hosts, a node writer adapter, and Uint8Array<->JSON stream adapters. It’s possible to write a `pump` function in `streams.js` that works on any (writer, reader) pair, which I’ve found use for in three places, connecting messaging hops. Chrome Extensions have their own stream primitive. The web has message ports, web sockets, and so on that all benefit from normalization.

I could certainly make this entire stack depend on running under hardened JavaScript, but it seems like a cheap missed opportunity to reach for broader JavaScript ecosystem adoption.

@mhofman brought up the issue of having a hard-soft-portable library initialize before lockdown and get used after lockdown. That is a hazard that really only applies in the start compartment, which I think is a special enough case, wherein we generally advise that everything that can run after lockdown should. Once an Endo environment is bootstrapped, as it should be for any client programs, the concern doesn’t exist.

**Comment by kriskowal** [comment]
I’ve retracted the commit that relaxes the dependence on harden. All that should remain is the type rename and I’ve amended the issue description to reflect that scope. Also, following up with #4208.

**Review by erights — CHANGES_REQUESTED** [review]
LGTM, thanks.

Time to remove the `.domain` mitigation (which was always weak tea here anyway)?

**Review by erights — APPROVED** [review]

**Comment by erights** [comment]
PR description still talks about softening harden.

**Pull request merged**